src/ATokenNaive.sol

```diff
@@ -0,0 +1,357 @@
+  /// @param support The depositor's vote preferences in accordance with the `VoteType` enum.
+  function expressVote(uint256 proposalId, uint8 support) external {
+    uint256 weight = getPastDeposits(msg.sender, governor.proposalSnapshot(proposalId));
+    if (weight == 0) revert("no weight");
+
+    if (_proposalVotersHasVoted[proposalId][msg.sender]) revert("already voted");
+    _proposalVotersHasVoted[proposalId][msg.sender] = true;
+
+    if (support == uint8(VoteType.Against)) {
+      proposalVotes[proposalId].againstVotes += SafeCast.toUint128(weight);
+    } else if (support == uint8(VoteType.For)) {
+      proposalVotes[proposalId].forVotes += SafeCast.toUint128(weight);
+    } else if (support == uint8(VoteType.Abstain)) {
+      proposalVotes[proposalId].abstainVotes += SafeCast.toUint128(weight);
+    } else {
+      revert("invalid support value, must be included in VoteType enum");
+    }
+  }
+
+  /// @notice Causes this contract to cast a vote to the Governor for all the
+  /// tokens it currently holds.  Uses the sum of all depositor voting
+  /// expressions to decide how to split its voting weight. Can be called by
+  /// anyone, but _must_ be called within `CAST_VOTE_WINDOW` blocks before the
+  /// proposal deadline.
+  /// @param proposalId The ID of the proposal which the Pool will now vote on.
+  function castVote(uint256 proposalId) external {
+    if (internalVotingPeriodEnd(proposalId) > block.number) revert("cannot castVote yet");
+    uint8 unusedSupportParam = uint8(VoteType.Abstain);
+    ProposalVote memory _proposalVote = proposalVotes[proposalId];
+
+    uint256 _proposalSnapshotBlockNumber = governor.proposalSnapshot(proposalId);
+
+    // Use the snapshot of total deposits to determine total voting weight. We cannot
+    // use the proposalVote numbers alone, since some people with deposits at the
+    // snapshot might not have expressed votes.
+    uint256 _totalDepositWeightAtSnapshot = getPastTotalDeposits(_proposalSnapshotBlockNumber);
+
+    // We need 256 bits because of the multiplication we're about to do.
+    uint256 _votingWeightAtSnapshot = IVotingToken(address(_underlyingAsset)).getPastVotes(
+      address(this), _proposalSnapshotBlockNumber
+    );
+
+    //      forVotesRaw          forVotesScaled
+    // --------------------- = ---------------------
+    //     totalDeposits        deposits (@snapshot)
+    //
+    // forVotesScaled = forVotesRaw * deposits@snapshot / totalDeposits
+    uint128 _forVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.forVotes) / _totalDepositWeightAtSnapshot
+    );
+    uint128 _againstVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.againstVotes) / _totalDepositWeightAtSnapshot
+    );
+    uint128 _abstainVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.abstainVotes) / _totalDepositWeightAtSnapshot
+    );
```

Comment:
So there's an invariant here that `_forVotesToCast + _againstVotesToCast + _abstainVotesToCast == _votingWeightAtSnapshot`? If so, we should compute abstain by subtracting. Currently all of these multiplications may round down so perhaps you could end up with that invariant not holding? Not sure what would happen in this case.

It looks like `_countVoteFractional` enforces `uint256(forVotes) + againstVotes + abstainVotes <= uint128(weight)` so maybe I was wrong that you have to vote with full weight

Reply:
> If so, we should compute abstain by subtracting

Sure, that's a good point.

Reply:
Per above comment around not submitting full voting weight, this makes sense now and I agree, thanks